Save / Auto save now disabled only for project wide settings, but enabled for input assets (ISX-1556)

lyndon-unity opened the pull request:
### Description

Save / Auto save now disabled only for project wide settings, but enabled for input assets ([ISX-1556](https://jira.unity3d.com/browse/ISX-1556))

![image](https://github.com/Unity-Technologies/InputSystem/assets/33493311/9b9e161b-d90b-4c54-9ed0-81c85655db79)
![image](https://github.com/Unity-Technologies/InputSystem/assets/33493311/4237912b-7e38-4a6f-8858-b46656fed4cb)


### Changes made

Added explicit code to find save container and hide it (rather than being in the USS style sheet file) This is in the settings window code and not in the asset dialog code, so only effects project wide input actions.

UI TK editor now enabled for input assets as well as project wide input actions

Also removed some duplicate code for accessing stylesheets
And fixed some tests:

- a missing teardown call in InputUI tests
- a warning on an obsolete API (from 2023.2 onward)
- a camera override warning

### Notes

### Checklist

Before review:

- Changelog entry not added as this is not a user facing change from what they already see
- Tests changed, if applicable.
    - Functional tests - those listed above
- Docs for new/changed API's. - no API's changed

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

lyndon-unity committed:
Save / Auto save now disabled only for project wide settings, but enabled for input assets

Removed some duplicate code for accessing stylesheers

Fixed some tests:

a missing teardown call in InputUI tests
a warning on an obsolete API (from 2023.2 onward)
a camera override warning

lyndon-unity committed:
Corrected to disable save/auto save on project wide settings (rather than asset)

Removed disabling on style sheet side

Enabled UI TK editor to assets, both to support testing and for upcoming release.

lyndon-unity committed:
Update InputActionDrawerBase.cs

Corrected fix for obsolete function CanCacheInspectorGUI

jamesmcgill commented on Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/InputActionsEditorWindow.cs:
```diff
@@ -21,7 +21,7 @@ static EnableUITKEditor()
             // This is set to false to show the IMGUI version of the InputActionAsset Editor instead.
             // UITK Editor is always be used for the Project Settings Editor regardless of this setting.
-            InputSystem.settings.SetInternalFeatureFlag(InputFeatureNames.kUseUIToolkitEditorForAllAssets, false);
+            InputSystem.settings.SetInternalFeatureFlag(InputFeatureNames.kUseUIToolkitEditorForAllAssets, true);
```
Should we remove the feature flag now (and it it's associated code) to tidy that up?

Keeping the feature flag will make it easier to rollback if we are not 100% about it right now. But at some point this should be removed.

Pauliusd01 replied:
Can a new flag be added to instead switch to the IMGUI version? Let's say if there's an UITK issue in the future that's blocking some person it would be nice to offer that as a temporary workaround. We could remove it once UITK version has been battle tested properly with the users

jamesmcgill replied:
Yeah, that would be the proper way of keeping the option available (the reversed logic).

We really should not have `SetInternalFeatureFlag` anywhere in our code. That is an oversight that it was left in. The flag should be unset (`false`) for whatever the default state is. And then the flag should be set by external user code. By having it in our code it will possibly override what the user was trying to set this to.

lyndon-unity replied:
Good idea - flag renamed in latest comment, logic reversed to be opt in for emergency short term usage and SetInternalFeatureFlag no longer used in package

lyndon-unity replied:
I've also added a changelog

jamesmcgill reviewed (approved):
Nice one.

This will bee a user observable change though (going from pre1 to pre2) so it probably does warrant a changelog entry.

Run `format.ps1` to fix the formatting CI job.
The others failures are the known bee buildsystem issues.

Pauliusd01 reviewed (changes requested):
It appears that the save button doesn't actually save the asset properly (can't compare with current develop though since It's missing there)
https://github.com/Unity-Technologies/InputSystem/assets/54306142/593a57c3-3955-4765-90e5-a3d403f0b97c

jamesmcgill committed:
Merge branch 'develop' into save-buttons-only-asset

jamesmcgill reviewed (changes requested):
Hiding the auto-save button visually is not enough, as in this case auto-save needs to be logically enabled too.
Currently `InputEditorUserSettings.autoSaveInputActionAssets == false` and without the AutoSave button there is no way to set it to `true`.

Therefore `InputActionsEditorWindowUtils::SaveAsset()` never gets called now.

lyndon-unity committed:
Fixed formatting

lyndon-unity committed:
Renamed feature flag to opt into UIGUI editor for assets : USE_IMGUI_EDITOR_FOR_ASSETS

Renamed feature flag to opt into UIGUI editor for assets kUseIMGUIEditorForAssets (USE_IMGUI_EDITOR_FOR_ASSETS) and removed feature flag setting from package code.

Old kUseUIToolkitEditorForAllAssets (USE_UITOOLKIT_EDITOR) feature flag removed.

We didn't document the old flag and the new one is just for emergency so not documented either.

lyndon-unity commented:
> Nice one.
> 
> This will bee a user observable change though (going from pre1 to pre2) so it probably does warrant a changelog entry.
> 
> Run `format.ps1` to fix the formatting CI job. The others failures are the known bee buildsystem issues.

Done the format part, will update changelog too

lyndon-unity commented:
> It appears that the save button doesn't actually save the asset properly (can't compare with current develop though since It's missing there) https://github.com/Unity-Technologies/InputSystem/assets/54306142/593a57c3-3955-4765-90e5-a3d403f0b97c

I've noticed it does save, but it doesn't correctly update the dirty state.

jamesmcgill commented:
> InputActionsEditorWindowUtils::SaveAsset()

I didn't see `InputActionsEditorWindowUtils::SaveAsset()` getting called at all.
I see the file being changed when I close the editor, so I'm assuming AssetDatabase is flushing it then. But we cannot rely on this behaviour once we stop using `InputManager.asset` it will need to be manually serialized by that function.

lyndon-unity committed:
Fixed input asset saving to reset dirty state

Auto save in project wide actions no longer checks asset autosave flag

lyndon-unity commented:
> > InputActionsEditorWindowUtils::SaveAsset()
> 
> I didn't see `InputActionsEditorWindowUtils::SaveAsset()` getting called at all. I see the file being changed when I close the editor, so I'm assuming AssetDatabase is flushing it then. But we cannot rely on this behaviour once we stop using `InputManager.asset` it will need to be manually serialized by that function.

I think this is in reference to the project wide input asset when that asset level 'auto save' flag is unchecked.
This is an asset only flag and should not be checked in the project wide input asset flow. 
I've removed that in the latest commit.
(InputActionsEditorSettingsProvider.cs onStateChanged is where the save gets triggered of project wide input actions)

For asset save I've fixed the dirty state (and reset the cached asset data) when save occurs.
(InputActionsEditorWindow.cs onStateChanged is where auto save get triggered for assets)

lyndon-unity committed:
Edit button in inspector now opens new UI TK editor

lyndon-unity commented:
The latest commit makes the Edit asset button open the UI TK window too
![image](https://github.com/Unity-Technologies/InputSystem/assets/33493311/54c7deeb-8f51-493b-a3c4-eadc74b0c667)

lyndon-unity committed:
Formatting fix

Pauliusd01 reviewed (approved):
Looks good to me. Here's what I checked:

- Saving in UITK/IMGUI versions manually and with auto save (When renaming, deleting, adding things, etc)
- Undo functionality unchanged with both versions
- Project Wide Actions saving and undos also checked

lyndon-unity committed:
Merge branch 'develop' into save-buttons-only-asset
